docs/src/usage/distributed.rst

```diff
@@ -0,0 +1,162 @@
+see a list of the supported operations in the :ref:`API docs<distributed>`.
+
+.. note::
+   A lot of operations may not be supported or not as fast as they should be.
+   We are adding more and tuning the ones we have as we are figuring out the
+   best way to do distributed computing on Macs using MLX.
+
+Getting Started
+---------------
+
+MLX already comes with the ability to "talk" to MPI if it is installed on the
+machine. The minimal distributed program in MLX is as simple as:
+
+.. code:: python
+
+   import mlx.core as mx
+
+   x = mx.distributed.all_reduce_sum(mx.ones(10))
+   print(x)
+
+The program above computes sums the array ``mx.ones(10)`` across all
+distributed processes. If simply launched with ``python``, however, only one
+process is launched and no distributed communication takes place.
+
+To launch the program in distributed mode we need to use ``mpirun`` or
+``mpiexec`` depending on the MPI installation. The simplest possible way is the
+following:
+
+.. code:: shell
+
+    $ mpirun -np 2 python test.py
+    array([2, 2, 2, ..., 2, 2, 2], dtype=float32)
+    array([2, 2, 2, ..., 2, 2, 2], dtype=float32)
+
+The above launches two processes on the same (local) machine and we can see
+both standard output streams. The processes send the array of 1s to each other
+and compute the sum which is printed. Launching with ``mpirun -np 4 ...`` would
+print 4 etc.
+
+Installing MPI
+---------------
+
+MPI can be installed with homebrew, using the anaconda package manager or
+compiled from source. Most of our testing is done using ``openmpi`` installed
+with the anaconda package manager as follows:
+
+.. code:: shell
+
+    $ conda install openmpi
+
+Installing with homebrew may require specifying the location of ``libmpi.dyld``
+so that MLX can find it and load it at runtime. This can simply be achieved by
+passing the ``DYLD_LIBRARY_PATH`` environment variable to ``mpirun``.
+
+.. code:: shell
+
+    $ mpirun -np 2 -x DYLD_LIBRARY_PATH=/opt/homebrew/lib/ python test.py
+
+Setting up Remote Hosts
+-----------------------
+
+MPI can automatically connect to remote hosts and set up the communication over
+the network if the remote hosts can be accessed via ssh. A good checklist to
+debug connectivity issues is the following:
+
+* ``ssh hostname`` works from all machines to all machines without asking for
+  password or host confirmation
+* ``mpirun`` is accessible on all machines. You can call ``mpirun`` using its
+  full path to force a specific prefix on all machines.
+* Ensure that the ``hostname`` used by MPI is the one that you have configured
+  in the ``.ssh/config`` files on all machines. MPI could be using ``foo`` from
+  a ``hostname`` defined as ``foo.bar.com`` if the ``hostname`` of the current
+  machine is ``baz.bar.com``.
+
+An easy way to pass the host names to MPI is using a host file. A host file
+looks like the following:
+
+.. code::
+
+    host1 slots=1
+    host2 slots=1
```

Comment:
I'm slightly confused by the setup description here. Is `host1` supposed to be e.g. `foo` and `host2` supposed to be `baz` in this example? 

Maybe that third bullet could be simpler or switched to be an example of how to specify the host file and use it with MPI?

Reply:
That particular peculiarity cost me at least a day in debugging.

The host name in the host file can be whatever, ideally its FQDN. However, when the hosts share a part of the domain then MPI is not using the FQDN but just the distinct subset.

To make it concrete, if we have the following 2 hosts `foo.bar.com` and `baz.bar.com`, MPI will (may) not try to ssh to `foo.bar.com` but instead simply to `foo`. The problem that arises is that we may not have configured ssh to `foo` but only to `foo.bar.com`.

I removed it from the bullet and made it a note and added a few more clarifications on the hostfile example.